src/storage/mod.rs

```diff
@@ -72,16 +72,16 @@ impl<'a> Storage<'a> {
             DatabaseBackend::new(conn).into()
         }
     }
-    pub(crate) fn get(&self, path: &str) -> Result<Blob, Error> {
+    pub(crate) fn get(&mut self, path: &str) -> Result<Blob, Error> {
         match self {
             Self::Database(db) => db.get(path),
             Self::S3(s3) => s3.get(path),
         }
     }
 
-    fn store_batch(&mut self, batch: &[Blob], trans: &Transaction) -> Result<(), Error> {
+    fn store_batch(&mut self, batch: Vec<Blob>, trans: &Transaction) -> Result<(), Error> {
         match self {
-            Self::Database(db) => db.store_batch(batch, trans),
+            Self::Database(db) => db.store_batch(&batch, trans),
             Self::S3(s3) => s3.store_batch(batch),
         }
```

Comment:
What's the reason for making this an owned value? It was more flexible before.

Reply:
Owning the values allows retrying the failing uploads, otherwise there's all sorts of fun "Doesn't live long enough" errors

Reply:
It seems possible to do it without like we had it before .. can you post the error?

Reply:
The problem is that the batch members are moved into futures, so they're then required that lifetime. The alternative to this is cloning, this just makes it up-front to the caller instead of silently happening. Regardless, I don't think it matters, since there wasn't change to the callers past removing a reference

Reply:
Hmm it seems like there should be a way for rusoto to take a borrow instead of an owned value ... but that doesn't need to be part of this PR, I'll look into it later.

Reply:
The problem isn’t rusoto, it’s futures themselves. Even though we know the futures all live and die within the scope of the function, they still have the lifetime requirements that would be associated with a thread. In the same way that you can’t share a non-static reference across thread boundaries, you can’t share non-static references across future boundaries, because they could execute at completely arbitrary times, different threads, the works

Reply:
> In the same way that you can’t share a non-static reference across thread boundaries, you can’t share non-static references across future boundaries

`s/future/task/`

That only applies to futures spawned as top-level tasks, because the `spawn` function requires a `+ 'static` bound.